pilot/pkg/proxy/envoy/watcher.go

```diff
@@ -75,6 +72,29 @@ func (w *watcher) Run(ctx context.Context) {
 
 	go watchCerts(ctx, certDirs, watchFileEvents, defaultMinDelay, w.SendConfig)
 
+	shouldKickStart := true
+	if w.waitForCerts {
+		// If none of the cert/key files exist, should not kick start.
+		shouldKickStart = false
+		for _, cert := range w.certs {
+			for _, filename := range cert.Files {
+				if _, err := os.Stat(path.Join(cert.Directory, filename)); err == nil || !os.IsNotExist(err) {
+					shouldKickStart = true
+					break
+				}
+			}
+			if shouldKickStart {
+				break
+			}
+		}
+	}
+	if shouldKickStart {
+		log.Info("Try to start Envoy...")
+		w.SendConfig()
+	} else {
+		log.Info("Cert files not ready, wait until cert files exist to start Envoy.")
```

Comment:
I'm not familiar with the Pilot agent code or this code path. it look slike Watcher.Run() is only called once. What is suppose to happen if the else condition is run here?

Reply:
If "else" is executed, we won't start Envoy in this thread. We will instead use the watchCerts routine to start the Envoy when cert files are changed.

Reply:
I'm concerned that having multiple paths where we start Envoy is adding unnecessary complexity.  Perhaps it would be better to have the main thread wait for some trigger from this thread before starting Envoy the standard way?